terraform/modules/vault/acm.tf

```diff
@@ -0,0 +1,34 @@
+# The MIT License (MIT)
+#
+# Copyright (c) 2014-2019 Avant, Sean Lingren
+
+resource "aws_acm_certificate" "acm" {
+  count             = "${ var.route53_enabled ? 1 : 0 }"
+  domain_name       = "${ var.route53_domain_name }"
+  validation_method = "DNS"
+
+  tags = "${ merge(
+    map("Name", "${ var.name_prefix }_private"),
+    var.tags ) }"
+
+  lifecycle {
+    create_before_destroy = true
+  }
+
+  depends_on = ["aws_route53_record.www"]
+}
+
+resource "aws_route53_record" "dns_acm_validation" {
+  count   = "${ var.route53_enabled ? 1 : 0 }"
+  name    = "${ aws_acm_certificate.acm.domain_validation_options.0.resource_record_name }"
```

Comment:
I think that `domain_validation_options` can contain more than one record set, so the `count` should be equal to the length of `domain_validation_options`

Reply:
I tried to play around with `length()` and `[count.index]`, but Terraform failed with a type error. Also, [documentation](https://www.terraform.io/docs/providers/aws/r/acm_certificate_validation.html) shows separate resources for each alternate name as an example.
I would appreciate some help with this feature.

Reply:
Ya terraform has a hard time with this until the upcoming release. AWS will ask for a unique validation record for the specified `domain_name` and also for each specified `subject_alternative_names`. Since this module will not create SANs it should be fine to just use the `.0.` notation. For reference, I've worked around this in the past with something like this:

```terraform
resource "aws_acm_certificate" "cert" {
  domain_name = "${ var.fqdn }"

  validation_method = "DNS"

  subject_alternative_names = ["${ var.sans }"]

  tags = "${ merge(
    map("Name", "${ replace( var.fqdn, "*", "star" ) }"),
    map("fqdn", "${ replace( var.fqdn, "*", "star" ) }"),
    map("sans", "${ replace( join( " / ", var.sans ), "*", "star" ) }"),
    map("valid_domains", "${ replace( join( " / ", concat( list(var.fqdn), var.sans ) ), "*", "star" ) }"),
    var.tags )
  }"

  lifecycle {
    create_before_destroy = true
  }
}

resource "aws_route53_record" "cert" {
  count = "${ 1 + length( var.sans ) }"

  zone_id = "${ element( var.route53_zone_id, count.index ) }"
  type    = "${ lookup( aws_acm_certificate.cert.domain_validation_options[count.index], "resource_record_type" ) }"
  ttl     = 60
  name    = "${ lookup( aws_acm_certificate.cert.domain_validation_options[count.index], "resource_record_name" ) }"
  records = ["${ lookup( aws_acm_certificate.cert.domain_validation_options[count.index], "resource_record_value" ) }"]

  lifecycle {
    create_before_destroy = true
  }
}

resource "aws_acm_certificate_validation" "cert" {
  certificate_arn         = "${ aws_acm_certificate.cert.arn }"
  validation_record_fqdns = ["${ aws_route53_record.cert.*.fqdn }"]

  lifecycle {
    create_before_destroy = true
  }
}
```